File: frontend/src/segmentation/thermalColor.ts
/** Legacy green→red thermal fill + IoU fallback (port of thermal_color.py). */

import type { GeoJsonFc } from "../api";

export type ThermalIndicator =
  | "max_temperature"
  | "min_temperature"
  | "mean_temperature"
  | "median_temperature"
  | "std_temperature"
  | "var_temperature";

export const THERMAL_INDICATORS: { id: ThermalIndicator; label: string }[] = [
  { id: "max_temperature", label: "Max °C" },
  { id: "min_temperature", label: "Min °C" },
  { id: "mean_temperature", label: "Mean °C" },
  { id: "median_temperature", label: "Median °C" },
  { id: "std_temperature", label: "Std °C" },
  { id: "var_temperature", label: "Variance" },
];

export const LABELABLE_INDICATORS: ThermalIndicator[] = [
  "max_temperature",
  "mean_temperature",
  "median_temperature",
  "std_temperature",
];

export type ColorRange = { min: number; max: number };

export function getThermalColorForValue(
  value: number | null | undefined,
  minVal: number,
  maxVal: number,
  colorRange?: ColorRange | null,
): string {
  if (value == null || Number.isNaN(value)) return "#808080";
  if (maxVal <= minVal) return "#00FF00";

  let normalized: number;
  if (colorRange) {
    const effectiveMin = colorRange.min;
    const effectiveMax = colorRange.max;
    if (effectiveMax <= effectiveMin) return "#00FF00";
    if (value <= effectiveMin) return "#00FF00";
    if (value >= effectiveMax) return "#FF0000";
    normalized = (value - effectiveMin) / (effectiveMax - effectiveMin);
  } else {
    normalized = (value - minVal) / (maxVal - minVal);
  }
  normalized = Math.max(0, Math.min(1, normalized));
  const red = Math.round(normalized * 255);
  const green = Math.round((1 - normalized) * 255);
  return `#${red.toString(16).padStart(2, "0")}${green.toString(16).padStart(2, "0")}00`;
}

export function iouFallbackColor(iou: number | null | undefined): string {
  const v = Number(iou ?? 0);
  if (v > 0.5) return "#00FF00";
  if (v > 0.3) return "#FFFF00";
  return "#FFA500";
}

export function percentileRange(values: number[], lo = 85, hi = 95): ColorRange {
  const arr = values.filter((v) => Number.isFinite(v)).sort((a, b) => a - b);
  if (arr.length === 0) return { min: 0, max: 1 };
  if (arr.length === 1) return { min: arr[0] - 0.5, max: arr[0] + 0.5 };
  const pct = (p: number) => {
    const i = ((arr.length - 1) * p) / 100;
    const loI = Math.floor(i);
    const hiI = Math.ceil(i);
    if (loI === hiI) return arr[loI];
    return arr[loI] + (arr[hiI] - arr[loI]) * (i - loI);
  };
  let min = pct(lo);
  let max = pct(hi);
  if (max <= min) max = min + 1e-3;
  return { min, max };
}

export function collectIndicatorValues(
  features: { properties?: Record<string, unknown> | null }[],
  indicator: ThermalIndicator,
): number[] {
  const out: number[] = [];
  for (const f of features) {
    const v = f.properties?.[indicator];
    if (typeof v === "number" && Number.isFinite(v)) out.push(v);
  }
  return out;
}

export function colorizePairsGeojson(
  fc: GeoJsonFc,
  opts: {
    indicator: ThermalIndicator;
    thermalColoring: boolean;
    colorRange: ColorRange | null;
  },
): GeoJsonFc {
  const values = collectIndicatorValues(fc.features || [], opts.indicator);
  const globalMin = values.length ? Math.min(...values) : 0;
  const globalMax = values.length ? Math.max(...values) : 1;
  return {
    type: "FeatureCollection",
    features: (fc.features || []).map((f) => {
      const props = { ...(f.properties || {}) };
      const val = props[opts.indicator];
      const fill = opts.thermalColoring
        ? getThermalColorForValue(
            typeof val === "number" ? val : null,
            globalMin,
            globalMax,
            opts.colorRange,
          )
        : iouFallbackColor(props.iou as number | undefined);
      props.fill_color = fill;
      return {
        ...f,
        properties: props,
        geometry: f.geometry,
      };
    }),
  };
}

export function softLabel(value: number | null, green: number, red: number): number | null {
  if (value == null || Number.isNaN(value)) return null;
  if (value <= green) return 0;
  if (value >= red) return 1;
  if (red <= green) return 0;
  return (value - green) / (red - green);
}

File: backend/openpvscope/segmentation/thermal_color.py
"""Legacy thermal fill colors + soft labels (segmentation map / Save Labels)."""

from __future__ import annotations

from typing import Any

THERMAL_INDICATORS = (
    "max_temperature",
    "min_temperature",
    "mean_temperature",
    "median_temperature",
    "std_temperature",
    "var_temperature",
)

LABEL_INDICATORS = (
    "max_temperature",
    "mean_temperature",
    "median_temperature",
    "std_temperature",
)

GRAY = "#808080"


def get_thermal_color_for_value(
    value: float | None,
    min_val: float,
    max_val: float,
    color_range: dict[str, float] | None = None,
) -> str:
    """Green→red hex; clamp outside color_range (legacy get_thermal_color_for_value)."""
    if value is None:
        return GRAY
    try:
        v = float(value)
    except (TypeError, ValueError):
        return GRAY
    if max_val <= min_val:
        return "#00FF00"

    if color_range and isinstance(color_range, dict):
        effective_min = float(color_range.get("min", min_val))
        effective_max = float(color_range.get("max", max_val))
        if effective_max <= effective_min:
            return "#00FF00"
        if v <= effective_min:
            return "#00FF00"
        if v >= effective_max:
            return "#FF0000"
        normalized = (v - effective_min) / (effective_max - effective_min)
    else:
        normalized = (v - min_val) / (max_val - min_val)

    normalized = max(0.0, min(1.0, float(normalized)))
    red = int(normalized * 255)
    green = int((1.0 - normalized) * 255)
    return f"#{red:02x}{green:02x}00"


def iou_fallback_color(iou: float | None) -> str:
    """Legacy live path: lime / yellow / orange."""
    v = float(iou or 0.0)
    if v > 0.5:
        return "lime"
    if v > 0.3:
        return "yellow"
    return "orange"


def soft_label(value: float | None, green: float, red: float) -> float | None:
    """≤ green → 0, ≥ red → 1, else linear (legacy save_labels_for_indicator)."""
    if value is None:
        return None
    try:
        v = float(value)
    except (TypeError, ValueError):
        return None
    if v <= green:
        return 0.0
    if v >= red:
        return 1.0
    if red <= green:
        return 0.0
    return float((v - green) / (red - green))


def target_column_for_indicator(indicator: str) -> str:
    return f"{indicator.split('_')[0]}_t_target"


def percentile_range(values: list[float], lo: float = 85.0, hi: float = 95.0) -> dict[str, float]:
    """Auto range when indicator changes (legacy p85–p95)."""
    import numpy as np

    arr = np.asarray([float(v) for v in values if v is not None and np.isfinite(v)], dtype=np.float64)
    if arr.size == 0:
        return {"min": 0.0, "max": 1.0}
    if arr.size == 1:
        v = float(arr[0])
        return {"min": v - 0.5, "max": v + 0.5}
    p_lo = float(np.percentile(arr, lo))
    p_hi = float(np.percentile(arr, hi))
    if p_hi <= p_lo:
        p_hi = p_lo + 1e-3
    return {"min": p_lo, "max": p_hi}


def colorize_feature_props(
    props: dict[str, Any],
    *,
    indicator: str,
    thermal_coloring: bool,
    color_range: dict[str, float] | None,
    global_min: float,
    global_max: float,
) -> str:
    if not thermal_coloring:
        return iou_fallback_color(props.get("iou"))
    val = props.get(indicator)
    return get_thermal_color_for_value(val if val is not None else None, global_min, global_max, color_range)
